op-node/rollup/engine/events.go

```diff
@@ -234,6 +340,36 @@ func (d *EngDeriver) OnEvent(ev event.Event) bool {
 			"safeHead", x.Safe, "unsafe", x.Unsafe, "safe_timestamp", x.Safe.Time,
 			"unsafe_timestamp", x.Unsafe.Time)
 		d.emitter.Emit(EngineResetConfirmedEvent(x))
+	case PromoteUnsafeEvent:
+		// Backup unsafeHead when new block is not built on original unsafe head.
+		if d.ec.unsafeHead.Number >= x.Ref.Number {
+			d.ec.SetBackupUnsafeL2Head(d.ec.unsafeHead, false)
+		}
+		d.ec.SetUnsafeHead(x.Ref)
+		d.emitter.Emit(UnsafeUpdateEvent(x))
+	case UnsafeUpdateEvent:
```

Comment:
long term thought: would be nice to be able to register handlers into this switch statement instead of making it longer.

Reply:
yes, the Register pattern is already tracked here: https://github.com/ethereum-optimism/optimism/issues/11152